### ISSUE TITLE
When the path changes for Downloads (on Windows), PicoTorrent by default continues to download files in the old path

### ISSUE THREAD

**Issue by Rakleed** [picotorrent/picotorrent#830]
When the path changes for Downloads (on Windows), PicoTorrent by default continues to download files in the old path.

I originally had a way for downloads like this - `C:\Users\<user_name>\Downloads`, but then I changed it to `D:\Downloads`. But when I open any .torrent file or add a magnet link, by default PicoTorrent offers to upload the file to `C:\Users\<user_name>\Downloads`, although PicoTorrent should offer `D:\Downloads`.

Could you please fix this problem?

**Comment by vktr** [comment]
Thanks for reporting! I'll have a look.

**Comment by vktr** [comment]
I can't reproduce this issue in either 0.18 or the latest unreleased version 🤔 Can you share the reproduction steps?

**Comment by Rakleed** [comment]
Steps to reproduce the problem:
1. The path for the Downloads folder was set by default as `C:\Users\<user_name>\Downloads`
2. Download and install PicoTorrent
3. Open any .torrent file and make sure that by default PicoTorrent offers a path for downloading `C:\Users\<user_name>\Downloads`
4. Close PicoTorrent
5. Change the path for the Downloads folder to `D:\Downloads` (or any other) and agree that all previously downloaded files will be transferred to a new path
6. Restart a computer
7. Open any .torrent file in PicoTorrent (the default path will remain old - `C:\Users\<user_name>\Downloads`)

**Comment by vktr** [comment]
Between step 5 and 6, do you click OK and/or close PicoTorrent before restarting the computer? I clicked OK, closed PicoTorrent and it showd the new save path as default when adding torrents.

**Comment by Rakleed** [comment]
I closed PicoTorrent, made the default path for the Downloads folder (`C:\Users\<user_name>\Downloads`), clicked OK, and then clicked Yes, agreeing to transfer all previously downloaded files to a new location. Then I rebooted the computer without opening PicoTorrent, changed the path for the download folder to another (in my case `D:\Downloads`), clicked OK, and then again Yes. Rebooted the computer, opened the .torrent file, but PicoTorrent again offered me to download the old path (`C:\Users\<user_name>\Downloads`).

There is another option that this is due to the fact that I recently used the official AdGuard removal tool ([advanced method](https://kb.adguard.com/en/windows/installation#advanced)), which modifies the registry, and maybe it has broken something else (since I use it every time I start Microsoft Word 2019 a macro error appears and AdGuard developers confirmed that this is a bug in their uninstaller).

**Comment by vktr** [comment]
What version is this? There's no option in the latest version to relocate existing torrents 🤔 Can you show a screenshot of that dialog?

**Comment by Rakleed** [comment]
I am using version 0.18, with GitHub. And I talked about this dialog:
<details>
<summary>Screenshot</summary>

![image](https://user-images.githubusercontent.com/19418601/77946672-606e7280-72cb-11ea-9f78-8561d6eb9fa1.png)

</details>

**Comment by vktr** [comment]
That dialog is not something created by PicoTorrent... Did you download v0.18 from GitHub? Or is it someone else's fork?

**Comment by Rakleed** [comment]
I downloaded PicoTorrent from GitHub. So I didn’t seem to write anywhere that this dialog creates PicoTorrent. The system creates this dialog after changing the path and clicking OK.

**Comment by vktr** [comment]
Well, what application creates that dialog then?

**Comment by Rakleed** [comment]
Most likely Explorer or Windows itself.

**Comment by thec0mpler** [comment]
If you don't mind me, I think I can clarify.

In Windows there are 'default' folders. In this case for Downloads, Documents, Pictures, etc. You can change the location of these folder. This is was Rakleed did. Lets call them A (old) and B (new). When changing, Windows will create folder B and asks if you want to move the files from location A to B. No matter what you choose, PicoTorrent will not change anything.

Let my explain why I think that. PicoTorrent will set [default_save_path](https://github.com/picotorrent/picotorrent/blob/master/res/dbmigrations/20181208115732_insert_default_settings.sql#L3) on first startup. Windows does provide this location, in this case A. When you add a torrent, this location will be in the [path_history](https://github.com/picotorrent/picotorrent/blob/master/src/picotorrent/addtorrentdialog.cpp#L118), the first one (I do not know when it is added). When you change the default location of the Download folder to B, nothing will change because the path history is hard coded in the database. It does not look again for the 'default' download folder. The default_save_path is also hard coded (in the config).

To recap. In Windows you can change the location of default folders. PicoTorrent does only checks the location of the default download folder the first time it starts and saves it. PicoTorrent does not check the location of the default download folder anymore, but uses the saved location.

I think this is an interesting bug/question.

- How does PicoTorrent know when the location is changed? Every time when is starts. But what if you do it while running?
- How do you know if the (torrent) files are moved or not?

Let me know if this helps?

**Comment by vktr** [comment]
Oh, I get it now. Thanks @thec0mpler and now I understand your steps @Rakleed. I don't think I want to classify this as a bug, but more a feature request since PicoTorrent works as intended.

Perhaps we can classify save paths in PicoTorrent and tag them somehow if they correspond to a default folder, and let PicoTorrent keep track of their actual path if so. It might be a cool change. I haven't thought about this scenario before.

Thanks again for reporting @Rakleed and thanks for explaining @thec0mpler!
